### PR TITLE
Implement CTFrame get string range methods

### LINKED ISSUE #920: Implement CTFrame string range APIs
- [ ] CTFrameGetStringRange
- [ ] CTFrameGetVisibleStringRange

### PR DESCRIPTION
**Pull request by aballway**
Implements CTFrameGetStringRange and CTFrameGetVisibleStringRange, as well as unit tests and adding to the manual test app

Resolves #920

### CODE REVIEW

**Comment by rajsesh-msft** [comment]
>  @Status Stub

Interoperable

---
Refers to: Frameworks/CoreText/CTFrame.mm:36 in 6d5a66b. [](commit_id = 6d5a66bd35f05aac400be6877f4ef7b06ec5b0b9, deletion_comment = False)

File: Frameworks/CoreText/CTFrame.mm [hunk 1]
```diff
@@ -37,17 +37,25 @@ - (void)dealloc {
...
-    return StubReturn();
+    CFIndex count = 0;
+    for (_CTLine* line in [static_cast<_CTFrame*>(frame)->_lines objectEnumerator]) {
+        count += line->_strRange.length;
+    }
+    return CFRangeMake(0, count);
 }
 
 /**
  @Status Stub
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTFrame.mm]: >Stub [](start = 9, length = 4)

Interoperable

File: Frameworks/CoreText/CTFrame.mm [hunk 2]
```diff
@@ -37,17 +37,25 @@ - (void)dealloc {
  @Notes
 */
 CFRange CTFrameGetStringRange(CTFrameRef frame) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    CFIndex count = 0;
+    for (_CTLine* line in [static_cast<_CTFrame*>(frame)->_lines objectEnumerator]) {
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTFrame.mm]: >frame [](start = 50, length = 5)

Does reference platform expect non-null values here? we will crash..

File: tests/unittests/CoreText/CTFrameTests.mm [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    NSMutableAttributedString* string = [[[NSMutableAttributedString alloc] initWithString:str] autorelease];
+    [string addAttribute:NSForegroundColorAttributeName value:[UIColor redColor] range:wholeRange];
+    [string addAttribute:NSFontAttributeName value:font range:wholeRange];
+
+    return string;
+}
+
+TEST(CTFrame, GetStringRange) {
+    CFAttributedStringRef string = (__bridge CFAttributedStringRef)getAttributedString(@"");
+    CTFramesetterRef framesetter = CTFramesetterCreateWithAttributedString(string);
```
Comment (rajsesh-msft) [inline tests/unittests/CoreText/CTFrameTests.mm]: framesetter must be released after usage?

File: tests/unittests/CoreText/CTFrameTests.mm [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    CGMutablePathRef path = CGPathCreateMutable();
+    CGPathAddRect(path, NULL, CGRectMake(0, 0, 100, 50));
+    CTFrameRef frame = CTFramesetterCreateFrame(framesetter, CFRangeMake(0, 0), path, NULL);
```
Comment (rajsesh-msft) [inline tests/unittests/CoreText/CTFrameTests.mm]: >frame [](start = 15, length = 5)

frame must be released..

**Comment by rajsesh-msft** [comment]
:clock1:

File: tests/unittests/CoreText/CTFrameTests.mm [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+#import <CoreText/CoreText.h>
+#import <CoreFoundation/CFAttributedString.h>
+
+static NSAttributedString* getAttributedString(NSString* str) {
+    UIFontDescriptor* fontDescriptor = [UIFontDescriptor fontDescriptorWithName:@"Times New Roman" size:40];
+    UIFont* font = [UIFont fontWithDescriptor:fontDescriptor size:40];
+
+    NSRange wholeRange = NSMakeRange(0, str.length);
+    NSMutableAttributedString* string = [[[NSMutableAttributedString alloc] initWithString:str] autorelease];
+    [string addAttribute:NSForegroundColorAttributeName value:[UIColor redColor] range:wholeRange];
```
Comment (ms-jihua) [inline tests/unittests/CoreText/CTFrameTests.mm]: fyi, there is a convenience initializer, [NSAttributedString initWithString:attributes:] that iirc applies the attributes to the whole range.

**Comment by ms-jihua** [comment]
:shipit:

**Pull request merged**